Month8/step1

MkZhdanov opened the pull request.

MkZhdanov committed:
feat: feat: созданы actions и reducers для работы с Redux, дабавлены функции для работы с сервером

MkZhdanov committed:
feat: реворкинг BurgerIngredients для работы с Redux и DND

MkZhdanov committed:
feat: реворкинг BurgerConstructor для работы с Redux и DND

MkZhdanov committed:
feat: настройка модальных окон для работы с Redux

MkZhdanov committed:
fix: очищен data.js от неактульных данных

gennady-bars commented on TEMP.jsx:
```diff
@@ -1,19 +1,82 @@
+      setDataIngredientsState({ ...dataIngredientsState, loading: true });
+      try {
+        const response = await fetch(
+          "https://norma.nomoreparties.space/api/ingredients"
```
`'https://norma.nomoreparties.space/api` везде повторяется в запросах. Нужно вынести в константу `BASE_URL` и импортировать везде в коде проекта, добавляя в конце изменяющуюся часть (эндпоит). Это обычная практика, чтобы одним движением можно было **сменить адрес**, но не менять эндпоинты.

gennady-bars commented on TEMP.jsx:
```diff
@@ -1,19 +1,82 @@
+  };
+
+  useEffect(() => {
+    const fetchDataIngredient = async () => {
+      setDataIngredientsState({ ...dataIngredientsState, loading: true });
+      try {
```
Все **асинхронные** функции проекта нужно сделать в виде `thunk`-функций редакса и объявить их в папке `actions`

gennady-bars commented on TEMP.jsx:
```diff
@@ -1,19 +1,82 @@
+        if (!response.ok) {
+          throw new Error(`Ошибка HTTP: ${response.status}`);
+        }
+        const rez = await response.json();
```
```js
.then(res => {
                if (res.ok) {
                    return res.json();
                }
                return Promise.reject(`Ошибка ${res.status}`);
            })

```

У Вас повторяется код **проверки ответа** во всех запросах. Чтобы этого избежать, нужно сделать отдельную функцию `checkResponse`, которая будет проверять ответ от сервера:

```js
function checkResponse(res) {
        // тут проверка ответа
    }

```

Ее нужно поместить в папку `utils` и импортировать туда, где нужна проверка на `ok`.

и теперь можно удалить кучу ненужного повторного кода и заменить его этим:

```js
 .then(checkResponse)

```

Обратите внимание, что передается только **ссылка** на функцию. Не нужно ее вызывать. Она сама вызовется, так как в `then` нужно передавать именно **функцию**, а не **вызов** функции.

-----------------------------------------

**Можно лучше**

Можно сделать **универсальную** функцию запроса с **проверкой ответа**, чтобы не дублировать эту проверку в каждом запросе:

```js

  function request(url, options) {
    // принимает два аргумента: урл и объект опций, как и `fetch`
    return fetch(url, options).then(checkResponse)
  }

```

И теперь просто нужно заменить все `fetch` на `request` и убрать дублирование проверки на `ok`. Все остальное будет без изменений. Код станет чище

И даже можно поместить внутрь `baseUrl`, чтобы не дублировать его в каждом запросе. Тогда нужно будет передавать просто `endpoint` в вызов

gennady-bars commented on src/services/api.js:
```diff
@@ -0,0 +1,42 @@
+export const getFetchIngredientsRequest = async () => {
+  try {
+    const response = await fetch(
+      "https://norma.nomoreparties.space/api/ingredients"
+    );
+    if (!response.ok) {
+      throw new Error(`Ошибка HTTP: ${response.status}`);
+    }
+    const data = await response.json();
```
* `'https://norma.nomoreparties.space/api` везде повторяется в запросах. Нужно вынести в константу `BASE_URL` и импортировать везде в коде проекта, добавляя в конце изменяющуюся часть (эндпоит). Это обычная практика, чтобы одним движением можно было **сменить адрес**, но не менять эндпоинты.
* У Вас повторяется код **проверки ответа** во всех запросах. Чтобы этого избежать, нужно сделать отдельную функцию `checkResponse`, которая будет проверять ответ от сервера:

gennady-bars commented on src/services/api.js:
```diff
@@ -0,0 +1,42 @@
+export const postFetchBurgerRequest = async (ingredients) => {
+  try {
+    const response = await fetch(
+      "https://norma.nomoreparties.space/api/orders",
+      {
+        method: "POST",
+        headers: {
+          "Content-Type": "application/json",
+        },
+        body: JSON.stringify({
+          ingredients,
+        }),
+      }
+    );
+    if (!response.ok) {
+      throw new Error(`Ошибка HTTP: ${response.status}`);
+    }
+    const data = await response.json();
```
* `'https://norma.nomoreparties.space/api` везде повторяется в запросах. Нужно вынести в константу `BASE_URL` и импортировать везде в коде проекта, добавляя в конце изменяющуюся часть (эндпоит). Это обычная практика, чтобы одним движением можно было **сменить адрес**, но не менять эндпоинты.
* У Вас повторяется код **проверки ответа** во всех запросах. Чтобы этого избежать, нужно сделать отдельную функцию `checkResponse`, которая будет проверять ответ от сервера:

gennady-bars commented on src/services/actions/order-details.js:
```diff
@@ -0,0 +1,31 @@
+  dispatch(createOrderRequest());
+  try {
+    const order = await postFetchBurgerRequest(ingredients);
+    if (order && order.success) {
```
**Можно лучше**


Если интересно, то можно создать **универсальную** функцию запроса `request` , которая будет содержать в себе уже **базовый урл**, проверку на **ok** и проверку на `success`, чтобы не дублировать их в каждом запросе.

```js

// 1 раз объявляем базовый урл
export const BASE_URL = "https://norma.nomoreparties.space/api/";

// создаем функцию проверки ответа на `ok`
const checkResponse = (res) => {
  if (res.ok) {
    return res.json();
  }
  // не забываем выкидывать ошибку, чтобы она попала в `catch`
  return Promise.reject(`Ошибка ${res.status}`);
};

// создаем функцию проверки на `success`
const checkSuccess = (res) => {
  if (res && res.success) {
    return res;
  }
  // не забываем выкидывать ошибку, чтобы она попала в `catch`
  return Promise.reject(`Ответ не success: ${res}`);
};

// создаем универсальную фукнцию запроса с проверкой ответа и `success`
// В вызов приходит `endpoint`(часть урла, которая идет после базового) и опции
const request = (endpoint, options) => {
  // а также в ней базовый урл сразу прописывается, чтобы не дублировать в каждом запросе
  return fetch(`${BASE_URL}${endpoint}`, options)
    .then(checkResponse)
    .then(checkSuccess);
};

```


И теперь не нужно будет дублировать все это в каждом запросе. Нужно просто заменить в запросах все `fetch` на `request`, удалив лишние проверки и базовый урл. Все остальное останется без изменений

Код станет намного чище. Будет мелкий аналог `axios`

Пример запроса ингредиентов

```js
const getIngredients =  () => request("ingredients");

```

В `get`-запросах даже не нужно указывать 2й параметр. Он получится `undefined`, что без проблем для `fetch` (который внутри).

gennady-bars replied:
Теперь не нужно проверять `success`. Все уже делается внутри `request`

gennady-bars commented on src/components/burger-constructor/burger-components/burger-components.jsx:
```diff
@@ -1,29 +1,48 @@
+  const renderedIngredients = useMemo(() => {
+    if (burgerIngredients) {
+      return burgerIngredients.map((item, index) => (
+        <BurgerComponent iconVis={true} data={item} key={index} />
```
`key={index} `

Для атрибута `key` в конструкторе нужно использовать для ингредиента уникальный `uuid` (Вам нужно его генерировать во время **сброса в зону** `DnD`). Это можно делать с помощью библиотеки `uuid` https://www.npmjs.com/package/uuid , или Вам должны были показать, как это сделать, иначе некорректно может работать DND. А в `key` будет уже просто **key={ item.uniqueId }**

gennady-bars replied:
Нужно генерировать uuid при создании экшена (желательно, в **`action creator`**). И созданный uuid должен приходить в экшене в `payload` или как отдельное поле, или сразу как часть объекта ингредиента. Посоветуйтесь с наставником, пожалуйста, если не получится
```js

// вот это - `action creator`
export const addIngridient = (item) => {
    return {
        type: ADD_INGRIDIENT,
        payload: {
            ...item, // используем `spread`, чтобы поменять ссылку на объект. Таким образом `redux` обновит его в хранилище
           uniqueId: uuid4()  // и добавляем в объект новое поле, которое потом будет использовано в `key`
        }
    }
}


```

gennady-bars replied:
[Почему индекс не нужно использовать, можно прочитать тут]( https://habr.com/ru/companies/hh/articles/352150/)

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

К сожалению, работу придется отклонить без потери итерации, 


* При закрытии попапа по `Escape` приложение ломается https://disk.yandex.ru/i/pFIm8VyY88bcXA 
* По заданию нужно сделать логику **сортировки** ингредиентов внутри заказа (вверх и вниз).
* Не могу удалить ингредиенты из заказа


Еще заметил ошибки:



* `'https://norma.nomoreparties.space/api` везде повторяется в запросах. Нужно вынести в константу `BASE_URL` и импортировать везде в коде проекта, добавляя в конце изменяющуюся часть (эндпоит). Это обычная практика, чтобы одним движением можно было **сменить адрес**, но не менять эндпоинты.
* Все **асинхронные** функции проекта нужно сделать в виде `thunk`-функций редакса и объявить их в папке `actions`
* У Вас повторяется код **проверки ответа** во всех запросах. Чтобы этого избежать, нужно сделать отдельную функцию `checkResponse`, которая будет проверять ответ от сервера
* Для атрибута `key` в конструкторе нужно использовать для ингредиента уникальный `uuid` (Вам нужно его генерировать во время **сброса в зону** `DnD`). Это можно делать с помощью библиотеки `uuid` https://www.npmjs.com/package/uuid , или Вам должны были показать, как это сделать, иначе некорректно может работать DND. А в `key` будет уже просто **key={ item.uniqueId }**
* 



**Можно лучше**

* Желательно **очищать конструктор** после успешного получения номера заказа с сервера в блоке `then`, чтобы пользователь мог следующий заказ сделать, не удаляя старые ингредиенты
* лучше не давать нажать на кнопку `Оформить заказ`, пока пользователь не добавил булку в заказ.
* лучше отображать **лоадер** (загрузку) при ожидании ответа от сервера с номером заказа

MkZhdanov committed:
fix: фикс ошибки при закрытии попапа по Escape

MkZhdanov committed:
fix: фикс логики сортировки ингредиентов внутри заказа
и функции удаления ингридиентов из заказа, для ингридиентов создаются уникальные ключи

MkZhdanov committed:
refactor:  обращение к серверу теперь через универсальную функцию

gennady-bars commented on src/services/reducers/burger-constructor.js:
```diff
@@ -0,0 +1,84 @@
+      };
+    }
+    case ADD_INGREDIENT_BURGER: {
+      const key = generateKey();
```
`      const key = generateKey();
`

Редьюсер должен быть **чистой функцией**. Генерировать `uuid` в самом редьюсере **нельзя**. Это делает редьюсер нечистой функцией. Нужно генерировать uuid при создании экшена (желательно, в **`action creator`**). И созданный uuid должен приходить в экшене в `payload` или как отдельное поле, или сразу как часть объекта ингредиента. Посоветуйтесь с наставником, пожалуйста, если не получится
```js

// вот это - `action creator`
export const addIngridient = (item) => {
    return {
        type: ADD_INGRIDIENT,
        payload: {
            ...item, // используем `spread`, чтобы поменять ссылку на объект. Таким образом `redux` обновит его в хранилище
           uniqueId: uuid4()  // и добавляем в объект новое поле, которое потом будет использовано в `key`
        }
    }
}


```

gennady-bars commented on src/services/actions/burger-ingredients.js:
```diff
@@ -0,0 +1,33 @@
+  dispatch(fetchIngredientsRequest());
+  try {
+    const ingredients = await getFetchIngredientsRequest();
+    if (ingredients && ingredients.success) {
```
` if (ingredients && ingredients.success) {`

Теперь не нужно проверять `success`. Все уже делается внутри `request`

gennady-bars commented on src/components/burger-constructor/burger-components/burger-components.jsx:
```diff
@@ -1,29 +1,52 @@
+        <BurgerComponent
+          iconVis={true}
+          data={ingredient}
+          key={ingredient.key}
```
`          key={ingredient.key}
`

Отлично, что указываете **уникальный** `key` для ингредиентов в заказе. Это гарантирует 100% логику `DnD`

gennady-bars commented on src/services/actions/burger-constructor.js:
```diff
@@ -0,0 +1,29 @@
+  payload: bun,
+});
+
+export const addIngredientBurger = (ingredient) => ({
+  type: ADD_INGREDIENT_BURGER,
+  payload: ingredient,
+});
+
+export const removeIngredientBurger = (ingredientIndex) => ({
+  type: REMOVE_INGREDIENT_BURGER,
+  payload: ingredientIndex,
+});
+
+export const removeBunBurger = () => ({
```
Хорошие экшен-криейтеры сделали

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная:



* Хорошие экшен-криейтеры сделали
* Отлично, что указываете **уникальный** `key` для ингредиентов в заказе. Это гарантирует 100% логику `DnD`
* Хорошая логика `DnD` и редакса

но есть некоторые недочеты:




* Редьюсер должен быть **чистой функцией**. Генерировать `uuid` в самом редьюсере **нельзя**. Это делает редьюсер нечистой функцией. Нужно генерировать uuid при создании экшена (желательно, в **`action creator`**). И созданный uuid должен приходить в экшене в `payload` или как отдельное поле, или сразу как часть объекта ингредиента. Посоветуйтесь с наставником, пожалуйста, если не получится
* Не стоит выбирать булку за пользователя при загрузке сайта. Нужно оставить в этом месте текст: `Пожалуйста, перенесите сюда булку и ингредиенты для создания заказа` или что-то в этом роде.




**Можно лучше**

* Желательно **очищать конструктор** после успешного получения номера заказа с сервера в блоке `then`, чтобы пользователь мог следующий заказ сделать, не удаляя старые ингредиенты
* лучше не давать нажать на кнопку `Оформить заказ`, пока пользователь не добавил булку в заказ.
* лучше отображать **лоадер** (загрузку) при ожидании ответа от сервера с номером заказа

MkZhdanov committed:
fix: фикс редьюсера  и мелкие правки

MkZhdanov committed:
fix: больше нет заказа по умолчанию + фикс кнопки оформления заказа

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились. С Новым Годом)

Удачного дальнейшего обучения.

MkZhdanov committed:
style: добабление комментариев
